provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+	sender messageSender,
+	logger slog.Logger,
+	filesystem afero.Fs,
+	workDirectory string,
+	provisioner sdkproto.DRPCProvisionerClient,
+	updateInterval time.Duration,
+	forceCancelInterval time.Duration) jobRunner {
+	m := new(sync.Mutex)
+
+	// we need to create our contexts here in case a call to cancel() comes immediately.
+	logCtx := slog.With(context.Background(), slog.F("job_id", job.JobId))
+	forceStopContext, forceStopFunc := context.WithCancel(logCtx)
+	gracefulContext, cancelFunc := context.WithCancel(forceStopContext)
+
+	return &runner{
+		job:                 job,
+		sender:              sender,
+		logger:              logger,
+		filesystem:          filesystem,
+		workDirectory:       workDirectory,
+		provisioner:         provisioner,
+		updateInterval:      updateInterval,
+		forceCancelInterval: forceCancelInterval,
+		mutex:               m,
+		cond:                sync.NewCond(m),
+		done:                make(chan any),
+		okToSend:            true,
+		forceStopContext:    forceStopContext,
+		forceStopFunc:       forceStopFunc,
+		gracefulContext:     gracefulContext,
+		cancelFunc:          cancelFunc,
+	}
+}
+
+func (r *runner) start() {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	defer r.forceStopFunc()
+
+	// the idea here is to run two goroutines to work on the job: run and heartbeat, then use the `r.cond` to wait until
+	// the job is either complete or failed.  This function then sends the complete or failed message --- the exception
+	// to this is if something calls fail() on the runner; either something external, like the server getting closed,
+	// or the heartbeat goroutine timing out after attempting to gracefully cancel.  If something calls fail(), then
+	// the failure is sent on that goroutine on the context passed into fail(), and it marks okToSend false to signal
+	// us here that this function should not also send a terminal message.
+	go r.run()
+	go r.heartbeat()
+	for r.failedJob == nil && r.completedJob == nil {
+		r.cond.Wait()
+	}
+	if !r.okToSend {
+		// nothing else to do.
+		return
+	}
+	if r.failedJob != nil {
+		r.logger.Debug(r.forceStopContext, "sending FailedJob")
+		err := r.sender.failJob(r.forceStopContext, r.failedJob)
+		if err != nil {
+			r.logger.Error(r.forceStopContext, "send FailJob", slog.Error(err))
+		}
+		r.logger.Info(r.forceStopContext, "sent FailedJob")
+	} else {
+		r.logger.Debug(r.forceStopContext, "sending CompletedJob")
+		err := r.sender.completeJob(r.forceStopContext, r.completedJob)
+		if err != nil {
+			r.logger.Error(r.forceStopContext, "send CompletedJob", slog.Error(err))
+		}
+		r.logger.Info(r.forceStopContext, "sent CompletedJob")
+	}
+	close(r.done)
+	r.okToSend = false
+}
+
+// cancel initiates a cancel on the job, but allows it to keep running to do so gracefully.  Read from isDone() to
+// be notified when the job completes.
+func (r *runner) cancel() {
+	r.cancelFunc()
+}
+
+func (r *runner) isDone() <-chan any {
+	return r.done
+}
```

Comment:
I can rename it, but I don't want things outside the runner directly accessing its data.

Reply:
The `Server` doesn't get a reference to `*runner`, it gets `jobRunner`, which is an interface and can't directly access `*runner` data.

Reply:
I'm not sure I understand the intent of adding an interface but keeping it in the same package with unexported fields. Is this a pattern you've seen in the stdlib? I'd assume a new package would be created because it's essentially a new namespace.